### PR TITLE
libmach: initial API bindings for mach core

### PR DESCRIPTION
**Pull request by zack466**
This pull request includes:
1. C API bindings for Core Mach, exported from `src/bindings.zig`
2. Changes to `build.zig` in order to build `libmach`
3. Test programs located in `libmach/` to check if `libmach` works as expected

- [X] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### CODE REVIEW

**Review by slimsag — APPROVED** [review]
This looks like a great starting point, and is a clear improvement over what we have here today so I'm going to merge it. We can continue to iterate on this. I'll leave feedback inline on how we can improve it next

File: src/bindings.zig [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    libmach.core_callbacks.core_update = core_update;
+}
+
+pub export fn mach_run() void {
+    if (libmach.core_callbacks.core_init == null) {
+        std.debug.print("Did not provide a core_init callback\n", .{});
+        return;
+    }
+    if (libmach.core_callbacks.core_update == null) {
+        std.debug.print("Did not provide a core_update callback\n", .{});
```
Comment (slimsag) [inline src/bindings.zig:36]: Perhaps we should panic in these cases?

File: src/bindings.zig [hunk 2]
```diff
@@ -0,0 +1,52 @@
+const std = @import("std");
```
Comment (slimsag) [inline src/bindings.zig:1]: Two thoughts for improvement:

1. Should these bindings live here, or in `libmach` I wonder? The same question would go for other Mach libraries in the future, e.g. when we expose a C ABI for the ECS library, I wonder where it should live?

2. We should plan on changing this API to reflect what Ayush said in Matrix, something like this as our long-term goal (but may require some changes to make this possible):

```
int main(void) {
    mach_platform_init(init_fn, usr_ptr);
    while (...) {
        mach_platform_update(update_fn, resize_fn, usr_ptr);
    }
    mach_platform_deinit(deinit_fn, usr_ptr);
}
```

File: src/bindings.zig [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+const gpu = @import("gpu");
+const Core = @import("Core.zig");
+const libmach = @import("platform/libmach.zig");
+const native = @import("platform/native.zig");
+
+// Current Limitations:
+// 1. Currently, ecs seems to be using some weird compile-time type trickery, so I'm not exactly sure how
+// `engine` should be integrated into the C API
+// 2. Core might need to expose more state so more API functions can be exposed (for example, the WebGPU API)
```
Comment (slimsag) [inline src/bindings.zig:10]: Once https://github.com/hexops/mach/pull/403 lands we should end up being ABI-compatible with the `webgpu.h` API, and Zig's stage2 compiler could even emit `webgpu.h` for us based on this too I think. Should be possible to start writing Mach core / WebGPU apps using this.

File: libmach/test.lisp [hunk 4]
```diff
@@ -0,0 +1,64 @@
+;; Tests the behavior of `libmach` using Common Lisp's CFFI
```
Comment (slimsag) [inline libmach/test.lisp:1]: Just curious @zack466 which language do you actually plan to use this from? I've personally never used lisp, but would be interested in maintaining some official Go bindings potentially.

File: src/bindings.zig (hunk 2 above)
Comment (iddev5) [reply src/bindings.zig:1]: My opinion on (1) I think the bindings should live inside src/ maybe changing the name to libmach.zig to be more clear. The libmach directory currently doesn't serves any purpose, it just contains examples.

File: libmach/test.lisp (hunk 4 above)
Comment (zack466) [reply libmach/test.lisp:1]: Common Lisp, haha

CFFI code is dead simple in Lisp (as you can see in `test.lisp`), which is really nice. I was actually originally trying to work with bgfx, but it was a huge pain to compile and use properly, so I eventually gave up. With Mach, on the other hand, a simple `zig build` does the trick!

**Pull request merged**